x/storage/keeper/payment.go

```diff
@@ -12,68 +11,117 @@ import (
+	}
+	return nil
 }
 
-func (k Keeper) ChargeDeleteBucket(ctx sdk.Context, bucketInfo *storagetypes.BucketInfo) error {
-	bill, err := k.GetBucketBill(ctx, bucketInfo)
+func (k Keeper) UnChargeBucketReadFee(ctx sdk.Context, bucketInfo *storagetypes.BucketInfo,
+	internalBucketInfo *storagetypes.InternalBucketInfo) error {
+	if internalBucketInfo.TotalChargeSize > 0 {
+		return fmt.Errorf("unexpected total store charge size: %d", internalBucketInfo.TotalChargeSize)
+	}
+
+	bill, err := k.GetBucketReadBill(ctx, bucketInfo, internalBucketInfo)
 	if err != nil {
-		return err
+		return fmt.Errorf("uncharge bucket read fee failed, get bucket bill failed, bucket: %s, err: %s", bucketInfo.BucketName, err.Error())
 	}
 	if len(bill.Flows) == 0 {
 		return nil
 	}
-	// should only remain at most 2 flows: charged_read_quota fee and tax
-	if len(bill.Flows) > 2 {
-		panic(fmt.Sprintf("unexpected left flow number: %d", len(bill.Flows)))
-	}
-	bill.Flows = GetNegFlows(bill.Flows)
+	bill.Flows = getNegFlows(bill.Flows)
 	err = k.paymentKeeper.ApplyUserFlowsList(ctx, []types.UserFlows{bill})
-	return err
+	if err != nil {
+		ctx.Logger().Error("uncharge bucket read fee failed", "err", err.Error())
+		return err
+	}
+	return nil
 }
 
-func (k Keeper) UpdateBucketInfoAndCharge(ctx sdk.Context, bucketInfo *storagetypes.BucketInfo, newPaymentAddr string, newReadQuota uint64) error {
+func (k Keeper) GetBucketReadBill(ctx sdk.Context, bucketInfo *storagetypes.BucketInfo,
+	internalBucketInfo *storagetypes.InternalBucketInfo) (userFlows types.UserFlows, err error) {
+	userFlows.From = sdk.MustAccAddressFromHex(bucketInfo.PaymentAddress)
+	if bucketInfo.ChargedReadQuota == 0 {
+		return userFlows, nil
+	}
+	price, err := k.paymentKeeper.GetStoragePrice(ctx, types.StoragePriceParams{
+		PrimarySp: bucketInfo.PrimarySpId,
+		PriceTime: internalBucketInfo.PriceTime,
+	})
+	if err != nil {
+		return userFlows, fmt.Errorf("get storage price failed: %w", err)
+	}
+
+	gvgFamily, found := k.virtualGroupKeeper.GetGVGFamily(ctx, bucketInfo.PrimarySpId, bucketInfo.GlobalVirtualGroupFamilyId)
+	if !found {
+		return userFlows, fmt.Errorf("get GVG family failed: %d", bucketInfo.GlobalVirtualGroupFamilyId)
+	}
+
+	// primary sp total rate
+	primaryTotalFlowRate := price.ReadPrice.MulInt(sdkmath.NewIntFromUint64(bucketInfo.ChargedReadQuota)).TruncateInt()
+
+	if primaryTotalFlowRate.IsPositive() {
+		userFlows.Flows = append(userFlows.Flows, types.OutFlow{
+			ToAddress: gvgFamily.VirtualPaymentAddress,
+			Rate:      primaryTotalFlowRate,
+		})
+	}
+
+	versionedParams, err := k.paymentKeeper.GetVersionedParamsWithTs(ctx, internalBucketInfo.PriceTime)
+	if err != nil {
+		return userFlows, fmt.Errorf("failed to get validator tax rate: %w, time: %d", err, internalBucketInfo.PriceTime)
+	}
+	validatorTaxRate := versionedParams.ValidatorTaxRate.MulInt(primaryTotalFlowRate).TruncateInt()
+	if validatorTaxRate.IsPositive() {
+		userFlows.Flows = append(userFlows.Flows, types.OutFlow{
```

Comment:
shall not `primaryTotalFlowRate` sub this?

Reply:
I will suggest to simplify it, we dont build stream for validator tax, but when storage provider withdraw from the virtual account, we split 1% of it to validator account.

Reply:
> shall not `primaryTotalFlowRate` sub this?

You can think it as extra fee. Will not sub sp's flow rate.

Reply:
> I will suggest to simplify it, we dont build stream for validator tax, but when storage provider withdraw from the virtual account, we split 1% of it to validator account.

Currently, the `validator tax pool` is used in challenge module to reward attestation `submitter`. If we don't create the flow, if will be not clear when withdrawal from `validator tax pool`.